### PR TITLE
22 unit tests

### PR DESCRIPTION
**Pull request by pat-alt**
- applies formatting
- adds format check to CI
- adds a generall `attack` method and an in-place version `attack!` that can be dispatched over different attack types (currently these are functions, which we probably want to change)
- no longer relying on custom crossentropy loss function which involves a bit of hard coding
- removed some hard-coded rules from `PGD`, `FGSM` and `AutoPGD`, in particular the call on the input size

### CODE REVIEW

File: src/attacks/autopgd/autopgd.jl
Comment (pat-alt) [inline src/attacks/autopgd/autopgd.jl:1]: I think this needs some work, so I've for now left this mostly unchanged.

File: src/attacks/common_utils.jl
Comment (pat-alt) [inline src/attacks/common_utils.jl:1]: I don't think we want to rely on this function. As far as I understand, we would generally sue the same loss function for the attack as for backprob?

File: src/attacks/attacks.jl
Comment (pat-alt) [inline src/attacks/attacks.jl:1]: Here is where I've added the general function(s) for applying an attack @rithik83. I think that instead of dispatching on functions, we probably want to dispatch on concrete types instead.

File: src/attacks/fgsm/fgsm.jl [hunk 1]
```diff
@@ -7,11 +7,19 @@ using Flux: onehotbatch, onecold
 
 # White-box Fast Gradient Sign Method (FGSM) attack by Goodfellow et al. (arxiv.org/abs/1412.6572)
 # Code adapted from (github.com/jaypmorgan/Adversarial.jl)
-function FGSM(model, x, y; loss=cross_entropy_loss, ϵ = 0.3, min_label=0, max_label=9, clamp_range = (0, 1))
-    w, h, c = size(x)
```
Comment (pat-alt) [inline src/attacks/fgsm/fgsm.jl:11]: I have generalized this to ensure these attacks can also be used with non-image data (see test file)

File: src/attacks/fgsm/fgsm.jl [hunk 2]
```diff
@@ -7,11 +7,19 @@ using Flux: onehotbatch, onecold
...
-    grads = gradient(x_adv -> loss(model(x_adv), y; min_label=min_label, max_label=max_label)[1], x_adv)[1]
+function FGSM(
+    model,
+    x,
+    y;
+    loss = cross_entropy_loss,
+    ϵ = 0.3,
+    clamp_range = (0, 1),
+)
+    x_adv = reshape(x, size(x)..., 1)       # why is this necessary @rithik83?
```
Comment (pat-alt) [inline src/attacks/fgsm/fgsm.jl:18]: why do we need this reshape?

File: src/attacks/square/square_attack.jl
Comment (pat-alt) [inline src/attacks/square/square_attack.jl:1]: haven't really touched this, just formatting

File: src/training/adversarial_training.jl
Comment (pat-alt) [inline src/training/adversarial_training.jl:1]: I would say that it's fine to have this here for prototyping, but I would actually prefer to focus on delivering functionality here that can be easily integrated with conventional Flux workflows. Happy to discuss!

File: test/train_mlp.jl
Comment (pat-alt) [inline test/train_mlp.jl:1]: A very simple toy example for testing.

File: test/train_mlp.jl [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+            _model = deepcopy(model)
+
+            for data in train_set
+                # Unpack this element (for supervised training):
+                input, label = data
+
+                # Attack the input:
+                println("Attacking input: $input")
+                input = attack(attack_type, input, label, _model, loss)
```
Comment (pat-alt) [inline test/train_mlp.jl:36]: @rithik83 note how I pass on the existing loss function that's also used for backprob.

**Review by pat-alt — COMMENTED** [review]
I've left some comments here. Hopefully that gives you some sense of what type changes I think are necessary, but perhaps it's also a good idea to jump on a call.

I'll keep this here as is for you to review before making more changes to the package.

File: src/attacks/common_utils.jl
Comment (rithik83) [reply src/attacks/common_utils.jl:1]: You're right. That function is kind of 'legacy' from back when I did autopgd for a batch rather than individual. 

there i needed to save the CE loss for every datapoint so I made this so that with one call it would find CE for everything in the batch (calling logitcrossentropy aggregates it all as is done for batch gradient descent ig)

passing on the same backprob loss function now is good. but maybe we leave this for now (because if we want to revert to generating adv examples for a batch for parallelism this can be useful) - what do you think?

Comment (pat-alt) [reply src/attacks/common_utils.jl:1]: yeh happy to keep in the repo for now, but would try to avoid relying on it. I think we should be able to implement parallelism through broadcasting, relying solely on what Flux offers in terms of loss functions.

File: src/attacks/fgsm/fgsm.jl (hunk 2 above)
Comment (rithik83) [reply src/attacks/fgsm/fgsm.jl:18]: The model I used to test, accepted inputs of the shape (w, h, c, batch_size). This function (and all attacks) have an input of a single image/datapoint etc (shape: (w, h, c)). so to make it compatible with the model to be able to test and iterate, I reshaped.

How do you suggest we do it? passing inputs as a batch probably is better long-term (speedups and what not) but then we may have to alter the attack algorithms to be able to deal with it. 
or for now, pass an input as (w, h, c, 1) itself so it doesn't have to keep being reshaped

File: src/attacks/fgsm/fgsm.jl [hunk 4]
```diff
@@ -7,11 +7,19 @@ using Flux: onehotbatch, onecold
...
-    x_adv = deepcopy(x)
-    x_adv = reshape(x_adv, w, h, c, 1)
-    grads = gradient(x_adv -> loss(model(x_adv), y; min_label=min_label, max_label=max_label)[1], x_adv)[1]
+function FGSM(
+    model,
+    x,
+    y;
+    loss = cross_entropy_loss,
+    ϵ = 0.3,
+    clamp_range = (0, 1),
```
[the remark this answers is not in the document]
Comment (rithik83) [reply src/attacks/fgsm/fgsm.jl:16]: Yeah it was like that to be able to test on MNIST. ofc if we are doing a different kind of data we can just pass on a different clamp range which will instead be used, and that should be fine

File: src/training/adversarial_training.jl
Comment (rithik83) [reply src/training/adversarial_training.jl:1]: Do you mean the adv-training mechanism?
Yeah we can discuss on this one, I'm not too sure I follow - particularly wrt conventional flux workflows as I am not too familiar with them

File: test/train_mlp.jl (hunk 3 above)
Comment (rithik83) [reply test/train_mlp.jl:36]: Yeah I see. this is a lot nicer

actually used to have it this way too but switched over to custom one when i batched inputs. but indeed as you said we can rely on flux implementations when we get back to that

File: test/Manifest.toml
Comment (rithik83) [inline test/Manifest.toml:1]: manifest file is usually git-ignored, right?

File: src/attacks/common_utils.jl
Comment (rithik83) [reply src/attacks/common_utils.jl:1]: Should be good then! Basically all I want to avoid wrt this one, is aggregating CE losses for all examples in a batch. every example should have its own CE loss (as with autoPGD we use each example's CE loss or targeted DLR loss to decide whether to halve the step size eta)

File: src/attacks/autopgd/autopgd.jl
Comment (rithik83) [reply src/attacks/autopgd/autopgd.jl:1]: Agreed. I will have to come back to this soon

File: src/attacks/attacks.jl
Comment (rithik83) [reply src/attacks/attacks.jl:1]: Is it possible to also add the AutoPGD and Square to the available attacks? Afaik the impls work (or at least they produce some perturbation that makes sense) and they take in the same main inputs (model, x, y before the ';'). remaining in kwargs ofc.

File: src/attacks/common_utils.jl
Comment (pat-alt) [reply src/attacks/common_utils.jl:1]: I think what you have in mind can be achieved by just using the `agg` keyword in the standard `Flux.Losses` functions:

```julia
julia> Flux.Losses.logitcrossentropy(yhat[:,1:3],ytrain[:,1:3], agg=(x -> x))
1×3 Matrix{Float32}:
 0.523598  0.242503  0.454158
```

File: src/attacks/attacks.jl
Comment (pat-alt) [reply src/attacks/attacks.jl:1]: Yeh will have a go at this now

File: src/attacks/fgsm/fgsm.jl (hunk 2 above)
Comment (pat-alt) [reply src/attacks/fgsm/fgsm.jl:18]: I think we probably want to use broadcasting for that instead

File: src/training/adversarial_training.jl
Comment (pat-alt) [reply src/training/adversarial_training.jl:1]: What I mean is that we don't necessarily need to ship a function for training where we either impose too many restrictions (e.g. in the current version you train wrt to clean and attacked samples) or we need to add a whole lot of arguments to maintain flexibility. 

`Flux` itself has a `train!` convenience method for plain vanilla training, but the docs mostly go through complete examples with custom training loops, because most of the time the plain vanilla approach will not be flexible enough users.

So I think an idea could be to try and overload the [train!](https://fluxml.ai/Flux.jl/dev/reference/training/reference/#Flux.Optimise.train!-NTuple{4,%20Any}) method for adversarial training or just not ship a method for training at all and demonstrate through examples how adversarial training can be achieved (e.g. simple toy example for testing below).

File: test/Manifest.toml
Comment (pat-alt) [reply test/Manifest.toml:1]: True, definitely for packages, although this is for the `test/` environment. But according to [this](https://discourse.julialang.org/t/special-environment-for-testing/90306/7) probably better to add to gitignore.

File: src/attacks/square/square_attack.jl [hunk 5]
```diff
@@ -7,10 +7,21 @@ include("utils.jl")
...
+    verbose = false,
+    clamp_range = (0, 1),
+)
     Random.seed!(0)
     w, h, c = size(x)
-    n_features = c*h*w
+    n_features = c * h * w
 
     # Initialization (stripes of +/-ϵ)
     init_δ = rand(w, 1, c) .|> x -> x < 0.5 ? -ϵ : ϵ
```
Comment (pat-alt) [inline src/attacks/square/square_attack.jl:27]: This is very specific to images, not sure how to address so will not use for testing (cannot be easily generalized)

Comment (rithik83) [reply src/attacks/square/square_attack.jl:27]: Yeah, square's implementation seems tailored to images. I followed the original paper in my impl

File: src/training/adversarial_training.jl
Comment (rithik83) [reply src/training/adversarial_training.jl:1]: I think we should ship functionality for adversarial training, as it would make the package more well-rounded in terms of being a 'one-stop shop' for all things adversarial ML in Julia.

I will definitely have a look at it, what I have in mind is to allow PGD/FGSM training alone for now (pretty much most adversarial training does this, or variants of FGSM training to bridge the gap - note that PGD training is a lot better but also more expensive) and train on only attacked samples which is the correct way to do it.

What we can do is have a 'baseline' implementation of AT adhering to the above - here we allow the user to specify PGD or FGSM and set the iterations, step-size and epsilon. We can also have a demonstration as you said, to show how custom ways of AT can be achieved (say if someone wants to train on both, or re-train on attacked samples later etc)

Comment (rithik83) [reply src/training/adversarial_training.jl:1]: On second thought, you are right - from my own experiences AT can be done in different ways (for ex: my way above, gradually increasing the attack strength with epochs, training on clean + perturbed etc) and we can allow the user to decide. maybe add a few scripts as you mentioned.

**Review by rithik83 — APPROVED** [review]
Looks good to me, a lot of necessary changes.
